app-ios/tutanota/Sources/Contacts/ContactsSynchronization.swift

```diff
@@ -0,0 +1,346 @@
+    var identifiersToRemove = Set(contacts.values)
+    
+    // we now need to create a request to remove all contacts from the user that match an id in idsToRemove
+    // it is OK if we are missing some contacts, as they are likely already deleted
+    let store = CNContactStore()
+    let fetch = CNContactFetchRequest(keysToFetch: [CNContactIdentifierKey] as [CNKeyDescriptor])
+    let save = CNSaveRequest()
+    var canDelete = false
+    try store.enumerateContacts(with: fetch) { contact, stopPointer in
+      guard identifiersToRemove.remove(contact.identifier) != nil else { return }
+      save.delete(contact.mutableCopy() as! CNMutableContact)
+      canDelete = true
+      
+      // stop early; nothing left to remove
+      if identifiersToRemove.isEmpty {
+        stopPointer.initialize(to: true)
+      }
+    }
+    
+    // nothing to delete?
+    if !canDelete {
+      return
+    }
+    
+    try store.execute(save)
+    
+    var oldIdentifiers = getSavedLocalContactsIdentifierMappings(forUserId: userId)
+    for contact in contacts {
+      oldIdentifiers[contact.key] = nil
+    }
+    writeSavedLocalContactsIdentifierMappings(oldIdentifiers, forUserId: userId, merge: false)
+  }
+  
+  private func getContactsMappingsDictionary() -> [String:Any] {
+    return self.userPreferencesProvider.getDictionary(forKey: CONTACTS_MAPPINGS) ?? [:]
+  }
+  
+  private func writeContactsMappingsDictionary(_ dictionary: [String:Any]) {
+    self.userPreferencesProvider.setValue(dictionary, forKey: CONTACTS_MAPPINGS)
+  }
+  
+  private func getTutaContactsGroupIdentifier() -> String? {
+    return getContactsMappingsDictionary()[TUTA_GROUP_IDENTIFIER] as? String
+  }
+  
+  private func setTutaContactsGroupIdentifier(_ identifier: String) {
+    var mappings = getContactsMappingsDictionary()
+    mappings[TUTA_GROUP_IDENTIFIER] = identifier
+    writeContactsMappingsDictionary(mappings)
+  }
+  
+  private func getSavedLocalContactsIdentifierMappings(forUserId userId: String) -> TutaToNativeContactIdentifiers {
+    let mappings = getContactsMappingsDictionary()
+    let userContacts = (mappings[userId] ?? [:]) as! TutaToNativeContactIdentifiers
+    return userContacts
+  }
+  
+  private func writeSavedLocalContactsIdentifierMappings(_ contacts: TutaToNativeContactIdentifiers, forUserId userId: String, merge: Bool) {
+    var mappings = getContactsMappingsDictionary()
+    
+    if merge {
+      var currentMappings = (mappings[userId] ?? [:]) as! TutaToNativeContactIdentifiers
+      for contact in contacts {
+        currentMappings[contact.key] = contact.value
+      }
+      mappings[userId] = currentMappings
+    } else if contacts.isEmpty {
+      mappings[userId] = nil
+    } else {
+      mappings[userId] = contacts
+    }
+    writeContactsMappingsDictionary(mappings)
+  }
+}
+
+
+/// Defines a mutable contact and functionality for commiting the contact into a contact store
+private class NativeMutableContact {
```

Comment:
```suggestion
private class NativeMutableContact: Equatable,  Hashable {
```

It might look cleaner to implement the `Equatable` & `Hashable` protocols inside the `NativeMutableContact` class itself.

Reply:
fwiw it's totally valid to create an extension per protocol conformance, then you only have the code pertaining to a certain protocol in its own block. It is (was?)  sometimes slower to compile thouggh.

Reply:
I looked over some Swift code bases in my free time & creating extensions to conform a class to protocols (as the code is now) is definitely the norm. My comment should be discarded. :+1: